app/src/main/java/org/wikipedia/bridge/CommunicationBridge.kt

```diff
@@ -0,0 +1,190 @@
+        } else {
+            communicationBridgeListener.webView.loadUrl(ServiceFactory.getRestBasePath(pageTitle.wikiSite) +
+                    RestService.PAGE_HTML_ENDPOINT + UriUtil.encodeURL(pageTitle.prefixedText))
+        }
+    }
+
+    fun cleanup() {
+        pendingJSMessages.clear()
+        pendingEvals.clear()
+        eventListeners.clear()
+        incomingMessageHandler?.removeCallbacksAndMessages(null)
+        incomingMessageHandler = null
+        communicationBridgeListener.webView.removeJavascriptInterface("pcsClient")
+        // Explicitly load a blank page into the WebView, to stop playback of any media.
+        loadBlankPage()
+    }
+
+    fun addListener(type: String, listener: JSEventListener) {
+        if (eventListeners.containsKey(type)) {
+            eventListeners[type]!!.add(listener)
+        } else {
+            val listeners = ArrayList<JSEventListener>()
+            listeners.add(listener)
+            eventListeners[type] = listeners
+        }
+    }
+
+    fun execute(js: String) {
+        pendingJSMessages.add("javascript:$js")
+        flushMessages()
+    }
+
+    fun evaluate(js: String, callback: ValueCallback<String>) {
+        pendingEvals[js] = callback
+        flushMessages()
+    }
+
+    fun evaluateImmediate(js: String, callback: ValueCallback<String?>?) {
+        communicationBridgeListener.webView.evaluateJavascript(js, callback)
+    }
+
+    private fun flushMessages() {
+        if (!isPcsReady || !isMetadataReady) {
+            return
+        }
+        for (jsString in pendingJSMessages) {
+            communicationBridgeListener.webView.loadUrl(jsString)
+        }
+        pendingJSMessages.clear()
+        for (key in pendingEvals.keys) {
+            communicationBridgeListener.webView.evaluateJavascript(key, pendingEvals[key])
+        }
+        pendingEvals.clear()
+    }
+
+    private var incomingMessageHandler: Handler? = Handler(Looper.getMainLooper(), Handler.Callback { msg ->
+        val message = msg.obj as BridgeMessage
+        if (!eventListeners.containsKey(message.action)) {
+            L.e("No such message type registered: " + message.action)
+            return@Callback false
+        }
+        try {
+            val listeners: List<JSEventListener> = eventListeners[message.action]!!
+            for (listener in listeners) {
+                listener.onMessage(message.action!!, message.data)
+            }
+        } catch (e: Exception) {
+            e.printStackTrace()
+            L.logRemoteError(e)
+        }
+        false
+    })
+
+    private class CommunicatingChrome : WebChromeClient() {
+        override fun onConsoleMessage(consoleMessage: ConsoleMessage): Boolean {
+            L.d(consoleMessage.sourceId() + ":" + consoleMessage.lineNumber() + " - " + consoleMessage.message())
+            return true
+        }
+    }
+
+    private inner class PcsClientJavascriptInterface {
```

Comment:
Would it be good if we add `@Suppress("unused")` to prevent this class from being removed accidentally?

Reply:
But it is used. I'm not seeing any warnings related to "unused" here...

Reply:
I should probably specified that the methods `onReceiveMessage()` and `setupSettings` are tagged unused. But after I tapped some `JavascriptInterface` relate action on IDE and the unused warnings are gone.